### DISCUSSION TITLE
unable to import inside module

### DISCUSSION THREAD

**Discussion by dominiquegarmier** [nuxt/nuxt#9744] [Questions (Nuxt 2)] [answered]
Here is what my project looks like:

I have a utility variable `bar` in `/utils/bar`:
```typescript
const bar = "world";
export { bar };
```

and I have a module `foo` in `/modules`:
```typescript
// foo/index.ts

import { bar } from "~/utils/bar"

export default function fooModule() {
  console.log("hello", bar);
}
```
in `nuxt.config.js`:
```javascript
export default {
  ...
  modules: ["~/modules/foo"],
  ...
}
```

When I run `npm run dev` I get:
```
 FATAL  Cannot find module '~/utils/bar'                                                                  15:02:10
Require stack:
- /.../modules/foo/index.ts

  Require stack:
  - modules/foo/index.ts
  at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
  at Function.resolve (node:internal/modules/cjs/helpers:100:19)
  at p (node_modules/jiti/dist/jiti.js:1:52987)
  at v (node_modules/jiti/dist/jiti.js:1:53886)
  at modules/foo/index.ts:1:177
  at Resolver.v [as _require] (node_modules/jiti/dist/jiti.js:1:54801)
  at Resolver.requireModule (node_modules/@nuxt/core/dist/core.js:381:29)
  at ModuleContainer.addModule (node_modules/@nuxt/core/dist/core.js:182:38)
  at node_modules/@nuxt/utils/dist/utils.js:639:43
  at async ModuleContainer.ready (node_modules/@nuxt/core/dist/core.js:55:5)


   ╭──────────────────────────────────────────────────────────────────────────╮
   │                                                                          │
   │    Nuxt Fatal Error                                                      │
   │                                                                          │
   │   Error: Cannot find module '~/utils/bar'                                │
   │   Require stack:                                                         │
   │   - /.../modules/foo/index.ts                                            │
   │                                                                          │
   ╰──────────────────────────────────────────────────────────────────────────╯
```
If I change `foo/index.ts`:

```typescript
// import { bar } from "~/utils/bar";

export default function fooModule() {
  // console.log("hello", bar);
}
```

It all works fine and I'm even able to import `/utils/bar` in pages and it works...
```html
<!-- pages/index.vue -->

<template>
  <div>Hello {{ bar }}</div>
</template>

<script lang="ts">
import Vue from "vue";

import { bar } from "~/utils/bar";

export default Vue.extend({
  data() {
    return {
      bar: bar
    };
  },
  mounted() {
    console.log("bar =", bar);
  }
});
</script>
```

So it seems `/utils` is unavailable during the build process.

Is there a specific folder where files like `bar/index.ts` should be placed in so they can be imported during module init? Or is there an option to tell nuxt to make those files available during module init? Or is something else wrong?

**Comment by kissu** [comment]
Tell me if my latest comment is helping: https://stackoverflow.com/questions/68950299/unable-to-import-a-module-from-another-module-in-nuxt/68986043?noredirect=1#comment121973957_68986043

**Reply by dominiquegarmier** [reply]
No, still getting the same error. I believe `@` and `~` act the same way. At least in the way they are defined in `tsconfig.json`.

```json
{
"paths": {
      "~/*": [
        "./*"
      ],
      "@/*": [
        "./*"
      ]
    },
}
```

**Reply by kissu** [reply]
Hm, not sure about TS. This is how it works in JS but I don't know what this config is doing so far.

**Reply by dominiquegarmier** [reply]
I quickly started a new project with javascript and it gives me the same error. The config I shared is only part of `tsconfig.json` and basically replaces any `'@'` or `'~'` in import statements with the proper path when compiling to javascript.

**Reply by kissu** [reply]
Btw, s I told you `modules: ["~/modules/foo"],` is not something to use in your context. Of course, you can have a `modules` directory but this will have nothing to do with a **Nuxt module**.

Also I'm not sure what you're trying to achieve with `import { bar } from "~/plugins/bar";`, where is it coming from even?

If you want to have some basic JS logic, put it into a `/utils` directory and import it directly into your component.

**Reply by dominiquegarmier** [reply]
Sorry it was suppose to be `import { bar } from "~/utils/bar";`

In my real project the module `modules/foo` is from what I know an actual Nuxt module like `@nuxtjs/axios`, which then binds a plugin to the context. (I more or less copied the way axios binds `$axios`)

I use functions from `/utils` both in that module and in my components. Or atleast I would like to, for now I need duplicates of these functions in `/utils` and in `modules/foo` because `modules/foo` cannot import anything from outside.

**Reply by kissu** [reply]
Feel free to edit your initial question to fix the typo. ^^

There is no module named `foo`: https://modules.nuxtjs.org/?q=foo&orderBy=desc, is it a homemade one?

If you want to bind something to the context, you can use `$inject` as explained here: https://nuxtjs.org/docs/2.x/directory-structure/plugins#inject-in-root--context

**Reply by dominiquegarmier** [reply]
I wrote my own module following [this](https://nuxtjs.org/docs/2.x/directory-structure/modules#write-your-own-module) guide. From what I under stand a "module" is simply a function that is executed sometime during the build process.

Im my project I then add the plugin like so:
```typescript
//index.ts

const APIModule: Module<Options> = function(_modelOptions) {
    this.addPlugin({
        src: path.resolve(__dirname, "plugin.ts")
    });
};
```
That plugin then injects functionality into my context like so:
```typescript
// plugin.ts

const APIPlugin: Plugin = (context, inject) => {
    const api_connector_instance = new APIConnector(context);
    inject("api", api_connector_instance);
};

export default APIPlugin;
```

The issue is that `APIConnector` depends on stuff in `/utils` which for some reason is not accessible inside a module.

**Comment by dominiquegarmier** [comment, marked as answer]
I found [this](https://github.com/nuxt/nuxt.js/issues/4580#issuecomment-614737257) issue comment talking about the same problem.
